### DISCUSSION TITLE
Not alter headers for requests

### DISCUSSION THREAD

**Discussion by marcosvega91** [mswjs/msw#713] [General]
This is a follow up for #678.

When a request is intercepted by MSW a custom header `x-msw-request-id` is sent through request. This could cause some issues if the request is not handled, for example the server could have some role on `access-control-allow-headers` that could block the request. 

At the same time, we need a way to identify the request. Could be better if we move the identifier into a cookie instead ?

**Comment by kettanaito** [comment]
Worth mentioning that `x-msw-request-id` is relevant only when running MSW in Node.js. The purpose of `x-msw-request-id` is to associate a request with a potential mocked/actual responses through the [Life-cycle events](https://mswjs.io/docs/extensions/life-cycle-events).

> In the browser, the worker sends _all_ the responses back to the client, with those responses containing a custom appended `requestId` property in the message (no need to alter request headers):

https://github.com/mswjs/msw/blob/d1017627ce052e6e1382cff8146b1bc85b78a4b8/src/mockServiceWorker.js#L115-L136

If I understand this correctly, the issue has 2 areas to surface:

- When the request is performed as-is by the "interceptors" library. This can happen with both unhandled requests and those requests that have a handler but that handler returns nothing (explicit bypass). 
- When calling `ctx.fetch(req)`, since `req.headers` will contain the `x-msw-request-id` header attached here:

https://github.com/mswjs/msw/blob/d1017627ce052e6e1382cff8146b1bc85b78a4b8/src/node/createSetupServer.ts#L55-L57

The `ctx.fetch` scenario would also have the `x-msw-bypass` header appended that's designed to let the worker know which requests to bypass:

https://github.com/mswjs/msw/blob/d1017627ce052e6e1382cff8146b1bc85b78a4b8/src/context/fetch.ts#L12

I agree that request headers may not be the best way to store this internal information. 

> Could be better if we move the identifier into a cookie instead? 

That's an interesting idea, but Service Worker cannot access request cookies:

- [Why cannot Service Worker access cookies?](https://stackoverflow.com/questions/35692988/why-cant-serviceworker-access-cookies)
- https://github.com/WICG/cookie-store/issues/37
- https://github.com/w3c/ServiceWorker/issues/837

I find the `x-*` headers being the least invasive option we can take.

**Comment by gomorizsolt** [comment]
I've just run into this issue as well... the pre-flight request fails due to the `x-msw-request-id` header is being present in `Access-Control-Request-Headers`. Any idea on how to get around it without altering the `Access-Control-Allow-Headers` header?

**Reply by kettanaito** [reply]
Hey, @gomorizsolt. 

Currently, the `x-msw-request-id` header is always appended to the request headers:

https://github.com/mswjs/msw/blob/fe6751f76108047fb27b16a875497693da1eb66d/src/utils/request/parseIsomorphicRequest.ts#L16

We don't have an API to opt-out from this behavior, as it would effectively break the functionality of [Life-cycle events](https://mswjs.io/docs/extensions/life-cycle-events). We are looking for contributors to help us brainstorm this and, hopefully, migrate from altering request headers for the sake of tracking requests and their responses. If you have any input on this, please don't hesitate to share it in the comments below. Thanks.

**Comment by kettanaito** [comment]
While the `x-msw-reqest-id` was relevant only to Node.js interception, the header was set for _all_ environments, as it was set in the `parseIsomorphicRequest` that's used everywhere. That particular behavior is what caused the CORS issues with the header not being allowed.

Now when the interceptors library generates and sets `request.id` by itself, we no longer need any additional properties on the request to associate responses with the corresponding requests. I've removed the `x-msw-request-id` header entirely in #1024.
